core/controller/src/main/scala/whisk/core/loadBalancer/SharedDataService.scala

```diff
@@ -54,45 +63,53 @@ class SharedDataService(storageName: String) extends Actor with ActorLogging {
   override def preStart(): Unit = {
     replicator ! Subscribe(storage, self)
     node.subscribe(self, initialStateMode = InitialStateAsEvents, classOf[MemberEvent], classOf[UnreachableMember])
-    replicator ! Update(storage, PNCounterMap.empty[String], writeLocal)(_.remove(node, "0"))
+    replicator ! Update(storage, ORMap.empty[String, PNCounterMap[Int]], writeLocal)(_.remove(node, "0"))
+  }
+  override def postStop(): Unit = {
+    node.unsubscribe(self)
   }
-  override def postStop(): Unit = node.unsubscribe(self)
 
   /**
    * CRUD operations on the counter, process cluster member events for logging
    * @return
    */
   def receive = {
 
-    case (IncreaseCounter(key, increment)) =>
-      replicator ! Update(storage, PNCounterMap.empty[String], writeLocal)(_.increment(key, increment))
-
-    case (DecreaseCounter(key, decrement)) =>
-      replicator ! Update(storage, PNCounterMap[String], writeLocal)(_.decrement(key, decrement))
-
+    case (IncreaseCounter(key, instance, increment)) =>
+      replicator ! Update(storage, ORMap.empty[String, PNCounterMap[Int]], writeLocal)(m => {
+        m + (key, m.getOrElse(key, PNCounterMap[Int]()).increment(instance.toInt, increment))
+      })
+    case (DecreaseCounter(key, instance, decrement)) =>
+      replicator ! Update(storage, ORMap.empty[String, PNCounterMap[Int]], writeLocal)(m => {
+        m + (key, m.getOrElse(key, PNCounterMap[Int]()).decrement(instance.toInt, decrement))
+      })
     case GetMap =>
       replicator ! Get(storage, readLocal, request = Some((sender())))
-
     case MemberUp(member) =>
       logging.info(this, "Member is Up: " + member.address)
-
     case MemberRemoved(member, previousStatus) =>
       logging.warn(this, s"Member is Removed: ${member.address} after $previousStatus")
-
-    case c @ Changed(_) =>
+    case c @ Changed(e) =>
       logging.debug(this, "Current elements: " + c.get(storage))
+      val res = c.get(storage).entries.mapValues(_.entries.mapValues(_.toInt))
+      if (res.nonEmpty) {
+        res.values.foreach(_.values.foreach(i => {
+          require(i >= 0, s"values cannot be less than 0 ${res}")
```

Comment:
Do we really want an exception to be thrown here? What's the supervision strategy for that?